garage/tf/samplers/off_policy_vectorized_sampler.py

```diff
@@ -0,0 +1,155 @@
+            ]
+            self.vec_env = VecEnvExecutor(
+                envs=envs, max_path_length=self.algo.max_path_length)
+        self.env_spec = self.algo.env.spec
+
+    @overrides
+    def shutdown_worker(self):
+        self.vec_env.close()
+
+    @overrides
+    def obtain_samples(self, itr):
+        paths = []
+        obses = self.vec_env.reset()
+        dones = np.asarray([True] * self.vec_env.num_envs)
+        running_paths = [None] * self.vec_env.num_envs
+
+        policy = self.algo.policy
+        if self.algo.es:
+            self.algo.es.reset()
+
+        for rollout in range(self.algo.max_path_length):
+            policy.reset(dones)
+            if self.algo.replay_buffer_type == Buffer.HER:
+                obs = [obs["observation"] for obs in obses]
+                d_g = [obs["desired_goal"] for obs in obses]
+                a_g = [obs["achieved_goal"] for obs in obses]
+                input_obses = np.concatenate((obs, d_g), axis=-1)
+            else:
+                input_obses = obses
+            if self.algo.es:
+                actions, agent_infos = self.algo.es.get_actions(
+                    input_obses, self.algo.policy)
+            else:
+                actions, agent_infos = self.algo.policy.get_actions(
+                    input_obses)
+
+            next_obses, rewards, dones, env_infos = self.vec_env.step(actions)
+            agent_infos = tensor_utils.split_tensor_dict_list(agent_infos)
+            env_infos = tensor_utils.split_tensor_dict_list(env_infos)
+            if agent_infos is None:
+                agent_infos = [dict() for _ in range(self.vec_env.num_envs)]
+            if env_infos is None:
+                env_infos = [dict() for _ in range(self.vec_env.num_envs)]
+
+            if self.algo.replay_buffer_type == Buffer.REGULAR:
+                self.algo.replay_buffer.add_transition(
+                    observation=obses,
+                    action=actions,
+                    reward=rewards * self.algo.reward_scale,
+                    terminal=dones,
+                    next_observation=next_obses,
+                )
+            elif self.algo.replay_buffer_type == Buffer.HER:
+                info_dict = {
+                    "info_{}".format(key): []
+                    for key in env_infos[0].keys()
+                }
+                for env_info in env_infos:
+                    for key in env_info.keys():
+                        info_dict["info_{}".format(key)].append(
+                            env_info[key].reshape(1))
+                self.algo.replay_buffer.add_transition(
+                    observation=obs,
+                    action=actions,
+                    goal=d_g,
+                    achieved_goal=a_g,
+                    terminal=dones,
+                    **info_dict,
+                )
+
+            if self.algo.replay_buffer_type == Buffer.HER and (
+                    rollout == self.algo.max_path_length - 1):
```

Comment:
If I understand correctly, this only constructs Hindsight samples from the very end of the rollout.
However, didn't the HER paper show that it's better to generate hindsight samples throughout the rollout? Has this already been discussed somewhere?

Reply:
I don't think it's been discussed.

Reply:
> If I understand correctly, this only constructs Hindsight samples from the very end of the rollout.
Yes. But I don't think it makes a lot difference.
Here is the process of DDPG.
```
for episode in range(1, M):
    for rollout in range(1, T):
        replay_buffer.add_transition(obs, action, reward, goal)
        for train in range(1, N):
            transitions = replay_buffer.sample(buffer_batch_size)
```
If we are using HerReplayBuffer, then in `replay_buffer.sample()`, it constructs Hindsight samples by calling `HerReplayBuffer._her_sample_transitions(    def _her_sample_transitions(episode_batch, sample_batch_size):
)`. It randomly selects `sample_batch_size` samples, then constructs Hindsight samples based on `future` strategy.

Reply:
>However, didn't the HER paper show that it's better to generate hindsight samples throughout the rollout? Has this already been discussed somewhere?

Do you mean the algorithm pseudocode in the original paper?
```
for episode in range(1, M):
    for rollout in range(1, T):
        replay_buffer.add_transition(obs, action, reward, goal)
        sample a set of additional goals G:=S
        for g in G:
            repaly_buffer.add_transitions(obs, actions, reward, g)
```
S here is the sample function of the goal in current episode. In the original paper, there are three strategies to sample the goal: `future`, `episode`, `random`, `final`. Experiments show `future` is the best. 
The `future` strategy replays with k random states which come from the same episode as the transition being replayed and were observed *after* it. I don't think it's possible to do this process during s<sub>t</sub> since it needs the goals in s<sub>t+future_offset</sub>.